text/0000-pibd-messages.md

```diff
@@ -0,0 +1,224 @@
+
+Clients can signify support for segment requests by signaling it in their capabilities using a new bit flag: `PIBD_HIST = 1 << 4`.
+
+The reference implementation for this RFC can be found in [\[3\]](#references), [\[4\]](#references) and [\[5\]](#references).
+
+## MMR segments
+[mmr-segments]: #mmr-segments
+We define a (P)MMR **segment** at segment height `h` as a set of `N = 2^h` consecutive leaves, along with the necessary data to verify membership of the leaves in the original MMR.
+Each segment is consecutive with the previous and is identified by the tuple `(H, h, i)` where `H` is the block hash, `h` is the segment "height" and `i` is a zero-based segment index: segment `(H, h, i)` contains the non-pruned leaves with leaf position in the interval `[i*N, (i+1)*N)` of the MMR state at the block with hash `H`.
+The final segment is allowed to have less than `N` elements because the full MMR does not necessarily contain a multiple of `N` leaves.
+
+Concretely, a segment contains the following items:
+- The identifier tuple `(H, h, i)`.
+- List of MMR positions and corresponding list of intermediary hashes, sorted by MMR position in ascending order.
+At a minimum it should contain all intermediary hashes that are required in order to reconstruct the subtree root.
+These lists are ignored for non-prunable MMRs.
+- A list MMR positions and a corresponding list of leaves in the segment (leaf index range `[i * 2^h, (i+1) * 2^h)`), sorted by MMR position in ascending order.
+At a minimum it should contain all unpruned leaves and the direct sibling of every unpruned leaf.
+- [The segment Merkle proof](#segment-merkle-proof).
+
+After receiving a segment, clients should verify if its contents are contained in the MMR by performing the following two steps: 
+1) calculate the [root of the segment](#segment-root) (using the leaf data and intermediary hashes) and 
+2) reconstruct the full MMR root using the segment root and the [segment merkle proof](#segment-merkle-proof) and compare it to the root in the block header.
+
+Clients are allowed to request fully pruned segments.
+In case of a fully pruned segment the list of leaves is expected to be empty and the list of hashes to contain a single entry: the first unpruned parent of the segment root.
+The proof is expected to contain the required hashes to reconstruct the MMR root from the first unpruned parent onwards.
+
+Clients should only ever request or respond with segments for the latest and first-to-latest canonical block for which `height % TXHASHSET_ARCHIVE_INTERVAL== 0`.
+
+Clients are free to choose a segment size by setting the height within a certain range in their requests.
+This allowed range is set differently depending on the MMR contents.
+The following table has concrete values for the different MMR types and the allowed height range.
+The upper bound has been chosen such that a single segment roughly corresponds to the maximum size of a block (1,348,032 B).
```

Comment:
Segment sizes and prunable MMRs potentially complicate things.
For the output and rangeproof MMRs the early segments are likely to be relatively sparsely populated and for a given height will be smaller than these estimated sizes, potentially significantly smaller.

We could end up requesting many small (in bytes) output and rangeproof segments.
I'm not sure how much we care about this right now though.

Reply:
Right, the current height limits are based on completely filled segments. We could instead place a limit on the number of leaves in a segment. This means either iterating through the MMR twice (once to count, the second to actually create the segment) or alternatively we use to bitmap to count the number of unpruned leaves in the segment range. One complication here is that we cannot simply take the cardinality in the range but have to account for pruned leaves where the direct sibling is unpruned.